host-bin/startxiwi

```diff
@@ -0,0 +1,82 @@
+#!/bin/sh -e
+# Copyright (c) 2015 The crouton Authors. All rights reserved.
+# Use of this source code is governed by a BSD-style license that can be
+# found in the LICENSE file.
+
+set -e
+
+APPLICATION="${0##*/}"
+OPTS_ENTER=''
+OPTS_XIWI=''
+
+USAGE="$APPLICATION [options] chroot_app [parameters]
+
+Wraps enter-chroot to launch a window or tab in Chromium OS for any graphical application.
+Applications launched in this way show in independent windows or tabs.
+
+By default, it will use the primary user on the first xiwi-enabled chroot found.
+
+Enter '-h' for help with 'enter-chroot' and 'xiwi' options.
+"
+
+USAGE_HELP="Usage: enter-chroot [options]
+
+`enter-chroot -h 2>&1 | grep -e ' -[bckntu]'`
+
+    Options: -l, -x, and -X are deprecated for the xiwi application.
+
+
+Usage: xiwi [options] chroot_app [parameters]
+
+`enter-chroot -t xiwi xiwi -h 2>&1 | grep -e ' -[FT]' | sed 's/^/    /g'`
+
+    xiwi will normally close when the application returns. Some gui applications
+    fork before or during normal operation, which can confuse xiwi and cause it to
+    quit prematurely. If your application does not have a parameter that prevents
+    it from forking, and crouton is unable to automatically detect the fork, you can
+    use -f to prevent xiwi from quitting automatically.
+    xiwi will quit if you close the Chromium OS window when nothing is displayed.
+  
+    A default window manager will full-screen all windows, unless chroot_app begins
+    with 'start' or is 'xinit'. You can use -F to force the chroot_app full-screen.
+
+    You can cycle through multiple windows inside the
+    application via Ctrl-Alt-Tab/Ctrl-Alt-Shift-Tab, or close them via
+    Ctrl-Alt-Shift-Escape.  If chroot_app begins with 'start' but you still want to
+    use the default window manager, specify the full path of the application.
+"
+
+# Exits the script with exit code $1, spitting out message $@ to stderr
+error() {
+    local ecode="$1"
+    shift
+    echo "$*" 1>&2
+    exit "$ecode"
+}
+
+options='bc:hk:n:t:u:fFT'
+while getopts $options OPT; do
+  case ${OPT} in
+    b) OPTS_ENTER="$OPTS_ENTER -${OPT}";;
+    c|k|n|t|u) OPTS_ENTER="$OPTS_ENTER -${OPT} ${OPTARG}";;
+    h) error 2 "$USAGE_HELP";;
+    f|F|T) OPTS_XIWI="$OPTS_XIWI -${OPT}";;
+   \?) error 2 "$USAGE";;
+  esac
+done
+shift "$((OPTIND-1))"
+
+for param in "$OPTS_ENTER"; do
+  ENTER_OPTS="$ENTER_OPTS'`echo -n "$param" | sed "s/'/'\\\\\\''/g"`'"
+  ENTER_OPTS="`echo -n "$ENTER_OPTS" | sed -e "s/^ //" -e "s/'//g"`"
+done
+for param in "$OPTS_XIWI"; do
+  XIWI_OPTS="$XIWI_OPTS'`echo -n "$param" | sed "s/'/'\\\\\\''/g"`'"
+  XIWI_OPTS="`echo -n "$XIWI_OPTS" | sed -e "s/^ //" -e "s/'//g"`"
+done
+
+eval exec sh -e "`dirname "\`readlink -f "$0"\`"`/enter-chroot" -t xiwi $ENTER_OPTS \
+     exec xiwi $XIWI_OPTS "$@" \
```

Comment:
Not enough levels of indirection.  `eval` will expand the string once, then treat the result as if it were a line of shell.  So it will see your

```exec sh -e "`dirname "\`readlink -f "$0"\`"`/enter-chroot" -t xiwi $ENTER_OPTS exec xiwi $XIWI_OPTS "$@"```

convert it into

```exec sh -e path with spaces/to/enter-chroot -t xiwi -c path with spaces AND a random unclosed "quote exec xiwi -T these are actually five parameters all running together```

and then mayhem and/or invalid syntax errors ensue.  Instead:

```eval "exec sh -e '`dirname "\`readlink -f "$0"\`"`/enter-chroot' -t xiwi $ENTER_OPTS exec xiwi $XIWI_OPTS \"\$@\""```

might work, or if you're concerned about paths to enter-chroot including `'`'s, then:

```
enterchroot="`dirname "\`readlink -f "$0"\`"`/enter-chroot`"
eval "exec sh -e \"\$enterchroot\" -t xiwi $ENTER_OPTS exec xiwi $XIWI_OPTS \"\$@\""
```

The trick is that we're delaying expansion of any variable we don't want to be doubly-expanded.

...eval is scary